src/middlewared/middlewared/pytest/unit/utils/test_pam_tdb.py

```diff
@@ -0,0 +1,325 @@
+LEGACY_ENTRY_KEY = 'rtpz6u16l42XJJGy5KMJOVfkiQH7CyitaoplXy7TqFTmY7zHqaPXuA1ob07B9bcB'
+LEGACY_ENTRY_HASH = '$pbkdf2-sha256$29000$CyGktHYOwXgvBYDQOqc05g$nK1MMvVuPGHMvUENyR01qNsaZjgGmlt3k08CRuC4aTI'
+INVALID_HASH_TYPE = '$pbkdf2-canary$29000$CyGktHYOwXgvBYDQOqc05g$nK1MMvVuPGHMvUENyR01qNsaZjgGmlt3k08CRuC4aTI'
+INVALID_SALT = '$pbkdf2-sha256$29000$CyGktHYOwXgvBYDQOqc0*g$nK1MMvVuPGHMvUENyR01qNsaZjgGmlt3k08CRuC4aTI'
+INVALID_HASH = '$pbkdf2-sha256$29000$CyGktHYOwXgvBYDQOqc05g$nK1MMvVuPGHMvUENyR01qNsaZjgGmlt3k08CRuC4a*I'
+MISSING_SALT = '$pbkdf2-sha256$29000$$nK1MMvVuPGHMvUENyR01qNsaZjgGmlt3k08CRuC4aTI'
+MISSING_HASH = '$pbkdf2-sha256$29000$CyGktHYOwXgvBYDQOqc05g$'
+PAM_DIR = '/etc/pam.d'
+PAM_FILE = 'middleware-api-key'
+PAM_AUTH_LINE = 'auth  [success=1 default=die]      pam_tdb.so debug '
+PAM_FAIL_DELAY = 2  # This is minimum that pam_tdb will delay failed auth attempts
+
+PAM_FILE_REMAINING_CONTENTS = """
+@include common-auth-unix
+@include common-account
+password        required        pam_deny.so
+session required        pam_deny.so
+"""
+
+
+def write_tdb_file(
+    username: str,
+    hashlist: list[str],
+    expired: bool = False
+) -> int:
+    """
+    Generate a tdb file based on the specified parameters
+    The resulting TDB will have one entry for `username` and
+    a varying amount of hashes.
+
+    Although each hash supports a separate expiry, we are only
+    concerned in these tests in that works overall.
+    """
+
+    keys = []
+    idx = 0
+
+    for idx, hash in enumerate(hashlist):
+        keys.append(user_api_key.UserApiKey(
+            userhash=hash,
+            id=BASE_ID + idx,
+            expiry=EXPIRED_TS if expired else 0
+        ))
+
+    entry = user_api_key.PamTdbEntry(username=username, keys=keys)
+
+    user_api_key.flush_user_api_keys([entry])
+
+    return BASE_ID + idx
+
+
+def truncate_tdb_file(username: str) -> None:
+    """
+    Truncate tdb entry to make pascal string point off end of buffer
+    If this sets PAM_AUTH_ERR then we need to look closely to make
+    sure we don't have parser issues in pam_tdb.c
+    """
+    hdl = tdb.open(user_api_key.PAM_TDB_FILE)
+    try:
+        data = hdl.get(username.encode())
+        hdl.store(username.encode(), data[0:len(data)-5])
+    finally:
+        hdl.close()
+
+
+def make_tdb_garbage(username: str) -> None:
+    """ fill entry with non-api-key data """
+    hdl = tdb.open(user_api_key.PAM_TDB_FILE)
+    try:
+        data = hdl.get(username.encode())
+        hdl.store(username.encode(), b'meow')
+    finally:
+        hdl.close()
+
+
+def make_null_tdb_entry(username: str) -> None:
+    """ throw some nulls into the mix for fun """
+    hdl = tdb.open(user_api_key.PAM_TDB_FILE)
+    try:
+        data = hdl.get(username.encode())
+        hdl.store(username.encode(), b'\x00' * 128)
+    finally:
+        hdl.close()
+
+
+@contextmanager
+def pam_service(
+    file_name: str = PAM_FILE,
+    admin_user: str | None = None,
+) -> None:
```

Comment:
flake8: return type is not None.
Actual flake8 message: Return type of generator function must be compatible with "Generator[Any, Any, Any]"
  "Generator[Any, Any, Any]" is not assignable to "None"